### PR TITLE
Make `NetworkService` callable for `ChainSync`

### PR DESCRIPTION
**Pull request by altonen**
Introduce a middleware called `NetworkServiceProvider` which the `ChainSync` can use to communicate with `NetworkService`. `ChainSync` is given a `NetworkServiceHandle` which it uses to call `NetworkServiceProvider` which then dispatches the calls to `NetworkService` on behalf of `ChainSync`.

This change will allow `ChainSync` to disconnect and report peers and in the future it'll be possible to send requests and notifications through the `NetworkServiceProvider`.

`NetworkServiceProvider` is needed only until the `ChainSync` object has been removed from `Protocol`. After that, a normal `NetworkService` handle can be passed onto `ChainSync` and these changes can be deprecated.

### CODE REVIEW

File: client/network/sync/src/lib.rs [hunk 1]
```diff
@@ -271,6 +271,8 @@ pub struct ChainSync<B: BlockT, Client> {
 	gap_sync: Option<GapSync<B>>,
 	/// Channel for receiving service commands
 	service_rx: TracingUnboundedReceiver<ToServiceCommand<B>>,
+	/// Handle for communicating with `NetworkService`
+	_network_service: service::network::NetworkServiceHandle,
```
Comment (bkchr) [inline client/network/sync/src/lib.rs:275]: Why isn't that used directly?

**Review by bkchr — APPROVED** [review]

**Review by arkpar — APPROVED** [review]

File: client/network/sync/src/lib.rs (hunk 1 above)
Comment (altonen) [reply client/network/sync/src/lib.rs:275]: Just trying to find ways to split the work into more reviewable PRs. This PR introduced the provider and next PR which moves the block requests to `ChainSync` will use the `network_service`.

**Comment by altonen** [comment]
bot rebase

**Comment by altonen** [comment]
bot merge

**Pull request merged**